### ISSUE TITLE
An inlined callee's imported binding is captured by a same-named binding in the caller

### ISSUE THREAD

**Issue by rboudrouss** [rboudrouss/reactant-analyzer#141]
Surfaced while measuring #2, but not caused by it: **208 rows before that change, 212 after.**

*(Retitled: the original diagnosis — "a module-level import is reported as a missing dep" — was wrong. A resolved or unresolved import on its own is handled correctly. The trigger is a name collision across an inlining boundary.)*

## Minimal repro

```ts
// useThing.ts — `t` here is a module-level import, therefore constant
import { useCallback } from "react";
import { t } from "@lingui/core/macro";

export const useThing = (x: number) => {
  return useCallback(() => { console.log(t`boom`, x); }, [x]);
};
```

```tsx
// App.tsx — a DIFFERENT `t`, same name, and this one is a hook value
import { useThing } from "./useThing";
import { useLingui } from "@lingui/react";

export function App({ x }: { x: number }) {
  const { t } = useLingui();
  const cb = useThing(x);
  return <button onClick={cb}>{t`hi`}</button>;
}
```

```
warn  missing-deps  var:t  `t` is used in this callback but not in its deps
                           array, and its value may change between renders
```

False. The `t` that callback reads is `useThing.ts`'s module import, which cannot change between renders. It is being resolved against `App`'s `t` instead.

## Cause

When the callee's body is inlined, its **local** bindings shadow correctly — a `const cfg = {…}` in the callee keeps its meaning even when the caller also has a `cfg` (verified both ways). Its **imported** bindings do not appear to be recorded as bindings at all, so the name falls through to the caller's scope and picks up whatever is there.

## Real-world instance

Exactly this shape, in twenty:

- `useOpenAskAiPageInSidePanel.ts:6` — `import { t } from '@lingui/core/macro'`
- `AIChatNoMoreBillingCreditsBanner.tsx:22` — `const { t } = useLingui()`

208 distinct `missing-deps` rows on `t` across the corpus. Two files alone do not reproduce it; the chain runs through several hooks, which is why the reduction above was built from the shape rather than by trimming the original.

## Direction

Predominantly **false positives**: an import is already maximally stable, so capturing it with a caller's value can only make it look less stable, not more. A false negative would need the caller's binding to be *more* provable than the callee's import — possible in principle (an opaque import replaced by a known local), so this is not purely cosmetic, but the FP direction is what the corpus shows.

## Why it is worth more than its count

`missing-deps` is one of the larger clusters, and this is the finding a first-time user hand-checks and dismisses the tool over. The remedy it implies — "add `t` to the deps" — is wrong advice for an import, and following it would be a no-op at best.

## Acceptance

- [ ] The repro above is silent
- [ ] A callee's local `const` still shadows a same-named caller binding (regression guard, passes today)
- [ ] A caller's own use of its `t` is still analysed normally
- [ ] The 208 `t` rows are gone from the corpus, measured both sides

**Comment by rboudrouss** [comment]
## Root cause, pinned

`callee_rename_map` in `src/ir/splice.rs` builds the set of names to alpha-rename from **params plus `Stmt::Let` targets only**:

```rust
let mut bound: HashSet<Var> = params.iter().cloned().collect();
for block in cfg.blocks.values() {
    for stmt in &block.stmts {
        if let Stmt::Let { var, .. } = stmt { bound.insert(var.clone()); }
    }
}
```

An imported name is neither a param nor a `let`, so it is not renamed. It stays a free variable of the spliced body and then resolves against whatever the caller happens to have.

## Why that is enough to cause it

Two probes settle what the rule does with each kind of name:

| the callback reads | result |
|---|---|
| `t` imported in the component's **own** file | silent |
| `t` with no binding anywhere (free, unresolvable) | silent |
| `t` free in the callee, **and** the caller has `const { t } = useLingui()` | **`missing-deps` fires** |

So `missing-deps` only reports names it can tie to a component-scope binding. The capture is the entire mechanism: nothing is wrong with how imports are treated, the callee's name is simply being resolved in the wrong scope.

It also means the fix is confirmed to work before writing it — renaming the callee's free `t` to `t#salt` leaves it unresolvable in the caller, which row 2 shows is silent.

## The general statement

In JavaScript a callee's free variable resolves in the **callee's module scope**, never in the caller's locals. So capturing a caller binding is always wrong, not just for imports: any free variable of an inlined body that is not a param is mis-scoped the moment the caller has the same name.

That argues for the broader fix — isolate *every* free var of the callee, not only the import-bound ones — rather than special-casing imports. It is also why this wants its own measured commit: it changes name resolution for every inlined body, and the corpus will move.

## Suggested shape

1. Extend `callee_rename_map` to cover the callee's free variables, not just its bound ones.
2. Keep params exactly as they are — they are already renamed and bound to the caller's args, which is the one legitimate channel between the two scopes.
3. Guard both directions: a callee local shadowing a caller name (passes today, must keep passing), and the capture above (fails today).
4. Measure. `t` alone is 208 rows; the total change will be larger.

**Comment by rboudrouss** [comment]
Fixed in `0f8f0ab`. Corpus **1358 → 1317 (−41: 41 removed, 0 added)** on the pinned corpus — the fix only withdraws claims, it never adds one.

## Correcting my own diagnosis, again

The title still said "imported binding". That is too narrow. Excalidraw supplies the clearer witness, and it has no import in it:

```ts
// packages/excalidraw/hooks/useTextEditorFocus.ts
export const saveCaretPosition = (doc) => { … };            // line 17, a module const

const saveCaretPositionToState = useCallback(() => {
  const position = saveCaretPosition(ownerDocument);        // line 78
}, […]);

return { saveCaretPosition: saveCaretPositionToState };     // line 102  ← the trap
```

The hook **returns** its result under the same name, so every consumer writes `const { saveCaretPosition } = useTextEditorFocus()` and binds it. Inlining then let the consumer's binding capture the callee's module-level function. Any module binding does this, not just an import — which is why the fix targets free variables in general.

## The implementation trap

The first attempt changed nothing, and it is worth recording why: by splice time a `useCallback` in the callee is already a `HookEntry` with its own CFG, leaving only a marker in the body. The capturable names are not in `body_cfg` at all. The rename map has to be built from the body **and** its hooks' sub-bodies — those are what read `t`.

## Scope of the fix

Only *colliding* free names are renamed, exactly as reasoned above: `fetch`, `console` and sibling utilities are recognised by name downstream, so blanket renaming would have traded this false positive for a false negative.

Three guards, each gated by removal:

- the capture is silent
- a callee local still shadows a same-named caller binding
- the caller's own binding is still analysed

## What the 41 removals are

All `missing-deps`; 40 in twenty, 1 in excalidraw. Sampled to source: `t` (Lingui macro import), `getFieldMetadataItemByIdOrThrow` (module import), `saveCaretPosition` (module const). All three are constant across renders and never belonged in a deps array.

Note the count differs from the 208 rows quoted earlier: that was *rows*, this is *distinct locations*, and the corpus was re-pinned in between (#15). Details in `docs/precision-log.md`.
